File: cses/05_range_queries/2184_missing-coin-sum-queries/solution.py
import sys
from array import array


def main():
    data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    q = int(data[idx]); idx += 1

    x = data[idx:idx + n]
    idx += n
    x = [int(v) for v in x]

    # Nén tọa độ theo giá trị đồng xu.
    vals = sorted(set(x))
    m = len(vals)
    comp = {v: i for i, v in enumerate(vals)}

    # Persistent segment tree trên trục giá trị [0, m-1], lưu bằng mảng phẳng.
    # SUM giữ tổng TIỀN của cây con (có thể tới ~2e14 nên dùng int64 'q').
    maxn = n * (m.bit_length() + 2) + 10
    LC = array('i', bytes(4 * maxn))
    RC = array('i', bytes(4 * maxn))
    SUM = array('q', bytes(8 * maxn))

    cnt = 1  # node 0 là null (SUM=0, LC=RC=0)

    # roots[i] = version sau khi chèn i đồng đầu tiên; roots[0] = cây rỗng.
    roots = [0] * (n + 1)

    m1 = m - 1
    for i in range(1, n + 1):
        val = x[i - 1]
        pos = comp[val]
        prev = roots[i - 1]
        # Sinh một root mới, chỉ nhân bản các node trên đường đi tới lá pos.
        new_root = cnt; cnt += 1
        cur_new = new_root
        cur_old = prev
        lo = 0; hi = m1
        while lo < hi:
            SUM[cur_new] = SUM[cur_old] + val  # cộng tiền lên đường đi
            mid = (lo + hi) >> 1
            if pos <= mid:
                RC[cur_new] = RC[cur_old]  # nhánh phải dùng chung với version cũ
                child = cnt; cnt += 1
                LC[cur_new] = child
                cur_old = LC[cur_old]
                cur_new = child
                hi = mid
            else:
                LC[cur_new] = LC[cur_old]  # nhánh trái dùng chung với version cũ
                child = cnt; cnt += 1
                RC[cur_new] = child
                cur_old = RC[cur_old]
                cur_new = child
                lo = mid + 1
        SUM[cur_new] = SUM[cur_old] + val  # lá: tổng tiền của giá trị này
        roots[i] = new_root

    # Xử lý từng truy vấn bằng greedy trên hiệu hai version.
    out = []
    ap = out.append
    _SUM = SUM; _LC = LC; _RC = RC; _vals = vals; _roots = roots
    for _ in range(q):
        a = int(data[idx]); b = int(data[idx + 1]); idx += 2
        na0 = _roots[a - 1]
        nb0 = _roots[b]
        # R = mọi tổng trong [0, R] đều tạo được từ đoạn [a, b].
        R = 0
        # DFS in-order theo thứ tự giá trị tăng dần; s = SUM[nb] - SUM[na].
        stack = [(na0, nb0, 0, m1)]
        while stack:
            na, nb, lo, hi = stack.pop()
            s = _SUM[nb] - _SUM[na]
            if s == 0:
                continue  # đoạn giá trị này không có đồng xu nào
            if _vals[lo] > R + 1:
                break  # giá trị nhỏ nhất còn lại > R+1 -> dừng
            if _vals[hi] <= R + 1:
                R += s  # gộp cả cây con vào reach một lần
                continue
            mid = (lo + hi) >> 1
            # Đẩy phải trước, trái sau -> pop trái trước (giữ đúng in-order).
            stack.append((_RC[na], _RC[nb], mid + 1, hi))
            stack.append((_LC[na], _LC[nb], lo, mid))
        ap(R + 1)

    sys.stdout.buffer.write(("\n".join(map(str, out)) + "\n").encode())


main()

File: cses/05_range_queries/2184_missing-coin-sum-queries/solution.cpp
#include <algorithm>
#include <cstddef>
#include <iostream>
#include <vector>

// Một node của persistent segment tree: con trái/phải và tổng TIỀN của cây con.
struct Node {
    int left = 0;
    int right = 0;
    long long sum = 0;
};

// Persistent segment tree dựng trên trục GIÁ TRỊ đã nén; mỗi lần chèn tạo ra
// O(log m) node mới nên giữ lại được mọi version theo prefix vị trí.
class PersistentSegmentTree {
public:
    PersistentSegmentTree(const int value_count, const int item_count)
        : value_count_(value_count) {
        // Ước lượng số node tối đa = số lần chèn * chiều cao cây.
        int levels = 1;
        for (int width = 1; width < value_count_; width <<= 1) {
            ++levels;
        }
        nodes_.reserve(
            static_cast<std::size_t>(item_count) * levels + 1U
        );
        nodes_.push_back(Node{});  // node 0 là null (sum = 0)
    }

    // Chèn thêm một đồng xu trị giá value vào vị trí giá trị position, sinh ra
    // một root mới dựa trên root previous_root.
    int insert(const int previous_root, const int position,
               const long long value) {
        return insert(previous_root, 0, value_count_ - 1, position, value);
    }

    const Node& operator[](const int index) const {
        return nodes_[index];
    }

private:
    int value_count_;
    std::vector<Node> nodes_;

    int insert(const int previous, const int low, const int high,
               const int position, const long long value) {
        // Sao chép node cũ rồi cộng thêm tiền của đồng xu mới lên đường đi.
        const int current = static_cast<int>(nodes_.size());
        nodes_.push_back(nodes_[previous]);
        nodes_[current].sum += value;

        if (low == high) {
            return current;  // tới lá: giá trị này đã được cập nhật tổng tiền
        }

        // Chỉ tạo node mới trên nhánh chứa position, nhánh còn lại dùng chung.
        const int middle = low + (high - low) / 2;
        if (position <= middle) {
            nodes_[current].left = insert(
                nodes_[previous].left, low, middle, position, value
            );
        } else {
            nodes_[current].right = insert(
                nodes_[previous].right, middle + 1, high, position, value
            );
        }
        return current;
    }
};

// Một khung của DFS tường minh: cặp node (trước/sau) và khoảng giá trị phủ.
struct TraversalFrame {
    int before;
    int after;
    int low;
    int high;
};

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n, q;
    std::cin >> n >> q;

    std::vector<long long> coins(n + 1);
    std::vector<long long> values;
    values.reserve(n);
    for (int index = 1; index <= n; ++index) {
        std::cin >> coins[index];
        values.push_back(coins[index]);
    }

    // Nén tọa độ theo giá trị đồng xu.
    std::sort(values.begin(), values.end());
    values.erase(std::unique(values.begin(), values.end()), values.end());

    // Dựng cây theo prefix vị trí: roots[i] = đã chèn i đồng đầu tiên x_1..x_i.
    PersistentSegmentTree tree(static_cast<int>(values.size()), n);
    std::vector<int> roots(n + 1, 0);
    for (int index = 1; index <= n; ++index) {
        const int position = static_cast<int>(
            std::lower_bound(values.begin(), values.end(), coins[index]) -
            values.begin()
        );
        roots[index] = tree.insert(roots[index - 1], position, coins[index]);
    }

    std::vector<TraversalFrame> stack;
    stack.reserve(64);
    while (q-- > 0) {
        int first, last;
        std::cin >> first >> last;

        // reachable = R: mọi tổng trong [0, R] đều tạo được từ đoạn [first, last].
        long long reachable = 0;
        stack.clear();
        // Hiệu hai version root[last] - root[first-1] cho tổng tiền theo giá trị.
        stack.push_back({
            roots[first - 1], roots[last], 0,
            static_cast<int>(values.size()) - 1
        });

        // DFS in-order theo thứ tự giá trị tăng dần, chạy greedy trên đường đi.
        while (!stack.empty()) {
            const TraversalFrame frame = stack.back();
            stack.pop_back();

            const long long subtree_sum =
                tree[frame.after].sum - tree[frame.before].sum;
            if (subtree_sum == 0) {
                continue;  // đoạn giá trị này không có đồng xu nào
            }
            if (values[frame.low] > reachable + 1) {
                break;  // giá trị nhỏ nhất còn lại > R+1 -> R+1 không tạo được
            }
            if (values[frame.high] <= reachable + 1) {
                reachable += subtree_sum;  // gộp cả cây con vào reach một lần
                continue;
            }

            // Node bắc cầu qua ngưỡng R+1: đẩy phải trước, trái sau để pop trái
            // trước (giữ đúng thứ tự in-order tăng dần).
            const int middle = frame.low + (frame.high - frame.low) / 2;
            stack.push_back({
                tree[frame.before].right,
                tree[frame.after].right,
                middle + 1,
                frame.high
            });
            stack.push_back({
                tree[frame.before].left,
                tree[frame.after].left,
                frame.low,
                middle
            });
        }

        std::cout << reachable + 1 << '\n';
    }
    return 0;
}
